bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/scriptengine/InvocationInterceptingScriptEngineWithInvocableAndAutoCloseable.java

```diff
@@ -116,6 +113,10 @@ public Object invokeMethod(Object o, String s, Object... objects) throws ScriptE
             return afterInvocation(super.invokeMethod(o, s, objects));
         } catch (ScriptException se) {
             throw afterThrowsInvocation(se);
+        } catch (NoSuchMethodException e) { // Make sure to unlock on a NoSuchMethodException to avoid deadlocks
```

Comment:
As @jpg0 mentioned: if you catch `Throwable` we are sure that the lock is always removed. (Also below) Otherwise LGTM.

Reply:
And then rethrow it? The method expects a `return` statement, so I either need to return something or throw something. If I want to rethrow it, I need to add `throws Throwable`.

Reply:
@J-N-K
I've fixed this in https://github.com/openhab/openhab-addons/pull/13924/commits/250601585414c50e92fc33937ebf3656fffb4ec8, but since I cannot throw a general `Exception` from these methods since that would clash with `Invocable.invokeMethod` and `Invoke.invokeFunction`, I handle all exceptions (except NPEs because SAT fails then) that can be thrown from these two methods.

And, I always need to return a value.

Reply:
LGTM

Reply:
Can you please then add an approving review so we can ping @jlaur for merging?

I think we don't need to wait for @jpg0 since his concerns were addressed and the other changes are exception handling which is nothing GraalJSScriptEngine specific.

Reply:
FYI the typical pattern here is to:

- Catch either any `Exception` (or `Throwable` in some cases), and allow consumers to process any of them (not just specific ones as you have done)
- If the resulting exception is not in the throws list, wrap it in an `InvocationTargetException` before rethrowing

It's not a big deal because your implementation should work, but is a bit more complex than it needs to be at the cost of preventing some consumer options (and is a non-standard interception advice pattern).

Reply:
@jpg0
Throwing a `InvocationTargetException` doesn‘t work because this gives a compilation error:
```
Project org.openhab.automation.jsscripting: Compilation failure: Compilation failure: 
[ERROR] /home/florianh/media/data_m2/git/openhab-addons/bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/scriptengine/InvocationInterceptingScriptEngineWithInvocableAndAutoCloseable.java:[126,13] Unhandled exception type java.lang.reflect.InvocationTargetException
[ERROR] /home/florianh/media/data_m2/git/openhab-addons/bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/scriptengine/InvocationInterceptingScriptEngineWithInvocableAndAutoCloseable.java:[144,13] Unhandled exception type java.lang.reflect.InvocationTargetException
``` 

If I add `InvocationTargetException` to `throws`, this clashed with the methods from the `Invocable` interface.

@J-N-K
I can now also rethrow a NPE because I added PMD suppresions for that class. 

Now all exceptions that are thrown by the two methods from the `Invocable` interface are properly rethrown.

Reply:
Oops sorry, it's `UndeclaredThrowableException` that you want.

Reply:
@jpg0 
This works, thanks. See https://github.com/openhab/openhab-addons/pull/13924/commits/8a6834d96c27c72d99cbdf2011ecf789f5421768. 

If you are happy with the changes, can you please add an approving review or comment so that we can merge this?

Reply:
I think that you mixed up `afterThrowsInvocation` and `afterInvocation` in the catch block? You shouldn't need to cast the return type, as `afterThrowsInvocation` should accept and return `Exception`?

Reply:
No, I deliberately did it that way.

I use `afterThrowsInvocation` only for exceptions (i.e. `ScriptException`) that the `OpenhabGraalJSScriptEngine` should be informed about. (FYI `afterThrowsInvocation` only accepts `ScriptException` and before this PR was also only called on a `ScriptException`, but not on `NoSuchMethodException` etc.).

For the other exceptions that are not a `ScriptException`, I use `afterInvocation` because they shouldn‘t be handled by the `OpenhabGraalJSScriptEngine` because they are related to missuse of the methods of the `Invocable` interface.

Reply:
I would not advise doing it that way. You cannot enumerate all the exception types which may be thrown by the delegate, only the non `RuntimeException` / declared ones (and even that isn't guaranteed as it's only checked at compile time). Better is to allow the consumer to make the decision on what types of exceptions they care about.

Additionally the current way mixes up whether the consumer is handling an exception or a return value? There is no way to tell if an exception is being thrown or being returned.

I did not notice that the current `afterThrowsInvocation` deals solely with `ScriptException`. I don't believe that is the right way to go.

Reply:
So what do you recommend to do?
I‘d like to get this finished and merged before the code freeze on Friday and my time is limited.

Reply:
Sorry for the back and forth... This is what I would do:

```
public Object invokeMethod(Object o, String s, Object... objects) throws ScriptException {
        try {
             return afterInvocation(super.invokeMethod(o, s, objects));
         } catch (Exception e) {
             Exception rethrow = afterThrowsInvocation(e);

             if(rethrow instanceof ScriptException) {
                 throw ((ScriptException)rethrow);
             } else if (rethrow instanceof RuntimeException) {
                  throw ((RuntimeException)rethrow);
             } else {
                  throw new UndeclaredThrowableException(rethrow);
             }
        }
}

...


public Exception afterThrowsInvocation(Exception e) { ... }
```

Reply:
@jpg0 
See https://github.com/openhab/openhab-addons/pull/13924/commits/8a6834d96c27c72d99cbdf2011ecf789f5421768, I hope this is the final commit ;-)

Reply:
Almost :-/

You updated the exception type, but didn't make the switch from `afterInvocation` to `afterThrowsInvocation`.

Reply:
See https://github.com/openhab/openhab-addons/pull/13924/commits/3140f02a04cd7d4862bf9e27331aad461ee53a1a.
I think I linked the wrong commit.

Maybe just have a look at the full change: https://github.com/openhab/openhab-addons/pull/13924/files.

Reply:
Ah yes, I see now. Great! There are two other issues*, but maybe we should look at them another time; I don't want to hold this up any longer and they are not things that are hard to change later.

* the are other methods which just cast the return of `afterThrowsInvocation`to `ScriptException` rather than performing the type checks. These casts may fail. Best is probably to extract the type checks into a separate method which each of these methods use to extract the correct exception type to throw.
* ideally the overriden implementation of `afterThrowsInvocation` should call `super.afterThrowsInvocation(e)` to get the exception to return, rather than using `e` directly, in case the parent class is also doing something with it.

Either way, let's get this merged!

Reply:
I‘ve created issue #13950 as a reminder.